### ISSUE TITLE
AttributeError: 'tuple' object has no attribute 'grad'

### ISSUE THREAD

**Issue by desertfire** [pytorch/torchdynamo#1979]
```
PYTORCH_TEST_WITH_DYNAMO=1 python test/test_nn.py -k test_AdaptiveAvgPool1d_cuda --verbose
```
passes with graph break;

```
PYTORCH_TEST_WITH_DYNAMO=1 python test/test_nn.py -k test_Adaptive --verbose
```
fails with
```
test_AdaptiveAvgPool1d_cuda (__main__.TestNN) ... Traceback (most recent call last):
  File "<string>", line 2, in <lambda>
AttributeError: 'tuple' object has no attribute 'grad'
ERROR RUNNING GUARDS <graph break in _backward> /scratch/binbao/work/pytorch/test/test_nn.py:90
  ___guarded_code.valid and
  ___check_type_id(input.grad, 109098144) and
  ___check_tensors(input, input.grad)
NULL ERROR: /scratch/binbao/work/pytorch/torch/csrc/dynamo/eval_frame.c:251
```

**Comment by desertfire** [comment]
cc @voznesenskym

**Comment by yanboliang** [comment]
Just chat with @desertfire, since this blocks several core test enhancement, I'll take a look.

**Comment by yanboliang** [comment]
@desertfire I was trying to reproduce it separately, but was not successful so far. But the root cause is sort of clear, which is the ```grad.data``` call. We rewrite ```.data``` call with ```detach()``` as it's deprecated already(since PT 0.4), just for backward compatibility. There are two ways to workaround and unblock:
* Make ```.data``` unimplemented at https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/tensor.py#L153. Since this is only for backward compatibility, so it should be fine.
* Just remove the ```.data``` call in the [test case](https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L92).

Let me know which way do you preferred, I can send a PR then.

**Comment by ezyang** [comment]
To solve this more generally, we need a more sophisticated design for tensor match guards that can handle attribute access. The crux of the problem is that you have to perform the type tests for all of the parents before you do the type test on the child (e.g., you must test `grad` is a tensor before you test `grad.data`). Ordinarily this would happen automatically, but because tensors are bundled up to be tested in one go, they violate this constraint. One way to fix this is to separate the tensor tests into levels: `___check_tensors([grad]) and ___check_tensors([grad.data])` but maybe in more complicated situations this may not work. Annoying!

**Comment by desertfire** [comment]
> One way to fix this is to separate the tensor tests into levels: `___check_tensors([grad]) and ___check_tensors([grad.data])` but maybe in more complicated situations this may not work. Annoying!

Would pytree utils be helpful for this case?

**Comment by ezyang** [comment]
Not... really?

**Comment by mlazos** [comment]
> To solve this more generally, we need a more sophisticated design for tensor match guards that can handle attribute access. The crux of the problem is that you have to perform the type tests for all of the parents before you do the type test on the child (e.g., you must test `grad` is a tensor before you test `grad.data`). Ordinarily this would happen automatically, but because tensors are bundled up to be tested in one go, they violate this constraint. One way to fix this is to separate the tensor tests into levels: `___check_tensors([grad]) and ___check_tensors([grad.data])` but maybe in more complicated situations this may not work. Annoying!

The tensor checks are usually performed last no? Every time I've seen this error we were missing type guards because we failed to find tensors in a nested container, I don't think the order was a problem.

**Comment by ezyang** [comment]
Yes, but the error here is in the type id check on tensor.grad, which happens before we tested tensor is a tensor
